pkg/agent/core/state.go

```diff
@@ -103,12 +102,12 @@ type LogConfig struct {
 
 // State holds all of the necessary internal state for a VM in order to make scaling
 // decisions
-type State struct {
-	internal state
+type State[A AlgorithmState] struct {
+	internal state[A]
 }
 
 // one level of indirection below State so that the fields can be public, and JSON-serializable
-type state struct {
+type state[A AlgorithmState] struct {
```

Comment:
Is there a reason why this has to be a type parameter, compared with just having an interface-typed field?

Reply:
Yeah, otherwise callers of `(*core.State).UpdateAlgorithmState()` would have to downcast to the specific type in order to do the updates — possible, but I figured why both downcasting when we can just use the exact type

Reply:
Can you take a quick look at https://github.com/neondatabase/autoscaling/pull/1140? I wanted to try to implement the same change, but without introducing generics, so that the impact is minimal.

Reply:
Ah, I see, the idea there is that the algorithm doesn't store the metrics? So, the "manual CU" algorithm would have shared state that's updated externally?

Reply:
> Ah, I see, the idea there is that the algorithm doesn't store the metrics? So, the "manual CU" algorithm would have shared state that's updated externally?

Yes and yes. These are the consequences of doing the minimal possible changes.

Reply:
Ok, I see. My two cents: I think the changes here aren't _too_ bad (it's mostly in tests), and having shared mutable state makes me nervous (specifically: changing things out from underneath another object) — but maybe that's just a holdover from my Rust programming :sweat_smile: